core/server/proxy/src/main/java/alluxio/proxy/s3/CompleteMultipartUploadHandler.java

```diff
@@ -312,25 +273,189 @@ public CompleteMultipartUploadResult call() throws S3Exception {
             }
           }
         }
-        String entityTag = Hex.encodeHexString(md5.digest());
         // persist the ETag via xAttr
+        String entityTag = Hex.encodeHexString(md5.digest());
         // TODO(czhu): try to compute the ETag prior to creating the file to reduce total RPC RTT
-        S3RestUtils.setEntityTag(mUserFs, objectUri, entityTag);
+        S3RestUtils.setEntityTag(mUserFs, objectTempUri, entityTag);
+        // rename the temp file to the target object file path
+        AlluxioURI objectUri = new AlluxioURI(objectPath);
+        mUserFs.rename(objectTempUri, objectUri, RenamePOptions.newBuilder()
```

Comment:
I have a little advise here. If we have got the ETag of the temp object, why not compare the two object's ETag to know if they are exactly the same objects?

Reply:
u meant to say compare tempobj etag with obj etag during renaming on master?

Reply:
yeah, I think it's good to compare before or during renaming.

Reply:
understood, actually there are multiple cases:
1) upload id match , etag match -> good, we are the same upload
2) upload id match, etag unmatch -> u r trying to complete a same mp uploadid with diff data from someone ahead of u, in this case, aws s3 will return NoSuchUpload error 
3) upload id unmatch, etag match -> we should overwrite as the metadata such as  tagging could be different.   (?) this i'm not particularly sure or if its necessary, since when we upload to UFS we didnt care about metadata at all it seems.
Let me put this into next batch of fixes and we can discuss it out ok?